librocksdb_sys/crocksdb/c.cc

```diff
@@ -5109,4 +5115,78 @@ void ctitandb_options_set_blob_run_mode(ctitandb_options_t* options,
   options->rep.blob_run_mode = static_cast<TitanBlobRunMode>(mode);
 }
 
+/* TitanReadOptions */
+struct ctitandb_readoptions_t {
+  TitanReadOptions rep;
+};
+
+ctitandb_readoptions_t* ctitandb_readoptions_create() { 
+  return new ctitandb_readoptions_t; 
+}
+
+void ctitandb_readoptions_destroy(ctitandb_readoptions_t* opts) { 
+  delete opts; 
+}
+
+uint64_t ctitandb_readoptions_key_only(ctitandb_readoptions_t* opts) {
+  return opts->rep.key_only;
+}
+
+void ctitandb_readoptions_set_key_only(ctitandb_readoptions_t* opts,
+                                        bool v) {
+  opts->rep.key_only = v;
+}
+
+crocksdb_iterator_t* ctitandb_create_iterator(
+    crocksdb_t* db,
+    const crocksdb_readoptions_t* options,
+    const ctitandb_readoptions_t* titan_options) {
+  crocksdb_iterator_t* result = new crocksdb_iterator_t;
+  if (titan_options == nullptr) {
+    result->rep = db->rep->NewIterator(options->rep);
+  } else {
+    *(ReadOptions*)&titan_options->rep = options->rep;
+    result->rep = ((TitanDB*)db->rep)->NewIterator(titan_options->rep);
+  }
+  return result;
+}
+
+crocksdb_iterator_t* ctitandb_create_iterator_cf(
+    crocksdb_t* db,
+    const crocksdb_readoptions_t* options,
+    const ctitandb_readoptions_t* titan_options,
+    crocksdb_column_family_handle_t* column_family) {
+  crocksdb_iterator_t* result = new crocksdb_iterator_t;
+  *(ReadOptions*)&titan_options->rep = options->rep;
+  result->rep = db->rep->NewIterator(options->rep, column_family->rep);
+  return result;
+}
+
+void ctitandb_create_iterators(
+    crocksdb_t *db,
+    crocksdb_readoptions_t* options,
+    ctitandb_readoptions_t* titan_options,
+    crocksdb_column_family_handle_t** column_families,
+    crocksdb_iterator_t** iterators,
+    size_t size,
+    char** errptr) {
+  std::vector<ColumnFamilyHandle*> column_families_vec;
+  for (size_t i = 0; i < size; i++) {
+    column_families_vec.push_back(column_families[i]->rep);
+  }
+
+  std::vector<Iterator*> res;
+  *(ReadOptions*)&titan_options->rep = options->rep;
+  Status status = db->rep->NewIterators(titan_options->rep, column_families_vec, &res);
```

Comment:
It seems to me `key_only` option may not pass to Titan, since the `TitanReadOptions` is convert to `ReadOptions` first before pass to `NewIterators`, which may loss the titan specific part. Can you check? I think the logic is cleaner if we either
* store `titan_inner: *mut TitanDB` pointer inside the `DB` struct, or
* cast `crocksdb_t *db` to `TitanDB` first

Reply:
yes, I do it in that way for `create_iterator`, see https://github.com/pingcap/rust-rocksdb/pull/309/files#diff-733c27ed8699d3baa6cbff3ac91a464bR5149.
Just forgot to update for `create_iterators`😅

Reply:
`*(ReadOptions*)&titan_options->rep = options->rep;` this is not the convert, it just copy the specific part of `ReadOptions` to `TitanReadOptions`.